extensions/amp-ad-exit/0.1/config.js

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {user} from '../../../src/log';
+import {FilterType} from './filters/filter';
+import {isValidClickDelaySpec} from './filters/click-delay';
+
+/**
+ * @typedef {{
+ *   targets: !Object<string, !NavigationTargetConfig>,
+ *   filters: (!Object<string, !FilterConfig>|undefined),
+ *   transport: (!Object<TransportMode, boolean>|undefined)
+ * }}
+ */
+export let AmpAdExitConfig;
+
+/**
+ * @typedef {{
+ *   finalUrl: string,
+ *   trackingUrls: (!Array<string>|undefined),
+ *   vars: (Variables|undefined),
+ *   filters: (!Array<string>|undefined)
+ * }}
+ */
+export let NavigationTargetConfig;
+
+/**
+ * @typedef {!Object<string, {defaultValue: (string|number|boolean)}>}
+ */
+export let Variables;
+
+/**
+ * @typedef {{
+ *   type: !FilterType,
+ *   delay: number
+ * }}
+ */
+export let ClickDelayConfig;
+
+/** @typedef {!ClickDelayConfig} */
+export let FilterConfig;
+
+/** @enum {string} */
+export const TransportMode = {
+  BEACON: 'beacon',
+  IMAGE: 'image',
+};
+
+/**
+ * Checks whether the object conforms to the AmpAdExitConfig spec.
+ *
+ * @param {*} config The config to validate.
+ * @return {!./config.AmpAdExitConfig}
+ */
+export function assertConfig(config) {
+  user().assert(typeof config == 'object');
+  if (config.filters) {
+    assertFilters(config.filters);
+  } else {
+    config.filters = {};
+  }
+  if (config.transport) {
+    assertTransport(config.transport);
+  } else {
+    config.transport = {};
+  }
+  assertTargets(config.targets, config);
+  return /** @type {!AmpAdExitConfig} */ (config);
+}
+
+function assertTransport(transport) {
+  for (const t in transport) {
+    user().assert(t == TransportMode.BEACON || t == TransportMode.IMAGE,
+                  `Unknown transport option: '${t}'`);
+    user().assert(typeof transport[t] == 'boolean');
+  }
+}
+
+function assertFilters(filters) {
+  for (const name in filters) {
+    user().assert(typeof filters[name] == 'object',
+                  `Filter specification '%s' is malformed`, name);
+    const type = filters[name].type;
+    switch (type) {
+      case FilterType.CLICK_DELAY:
+        user().assert(
+            isValidClickDelaySpec(filters[name]),
```

Comment:
instead of checking here, can we do it in filter's constructor. so we keep relevant logic in the same place, no need to jump around for new changes.